beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1725,13 +1733,98 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         result
     }
 
+    fn prune_abandoned_forks(
+        &self,
+        old_finalized_block_hash: SignedBeaconBlockHash,
+        new_finalized_block_hash: SignedBeaconBlockHash,
+        new_finalized_slot: Slot,
+    ) -> Result<(), Error> {
+        let mut abandoned_blocks: HashSet<(SignedBeaconBlockHash, BeaconStateHash, Slot)> =
+            HashSet::new();
+
+        let old_finalized_slot = self
+            .get_block(&old_finalized_block_hash.into())?
+            .ok_or_else(|| Error::MissingBeaconBlock(old_finalized_block_hash.into()))?
+            .message
+            .slot;
+
+        for (head_hash, _) in self.heads() {
+            let mut potentially_abandoned_blocks: HashSet<(
+                SignedBeaconBlockHash,
+                BeaconStateHash,
+                Slot,
+            )> = HashSet::new();
+
+            for (block_hash_, signed_beacon_block) in
+                ParentRootBlockIterator::new(&*self.store, head_hash)
+            {
+                let block_hash: SignedBeaconBlockHash = block_hash_.into();
+
+                if signed_beacon_block.message.slot > new_finalized_slot {
+                    // Tentatively schedule current block for deletion since it isn't know yet
+                    // whether the head is valid or not.
+                    potentially_abandoned_blocks.insert((
+                        block_hash,
+                        signed_beacon_block.message.state_root.into(),
+                        signed_beacon_block.message.slot,
+                    ));
+                } else if signed_beacon_block.message.slot == new_finalized_slot {
+                    if block_hash == new_finalized_block_hash {
+                        // The current head includes the newly finalized block, so we're either on
+                        // the canonical chain or on a legitimate candidate for the canonical
+                        // chain.  Do not remove any blocks.
+                        potentially_abandoned_blocks.clear();
+                    } else {
+                        // We've got a head which isn't based on the newly finalized block.  Delete
+                        // it and all its blocks.
+                        self.head_tracker.remove_head(head_hash);
+                        abandoned_blocks.extend(potentially_abandoned_blocks);
```

Comment:
We need the `new_finalized_blocks` here so that we can check how far back this fork started. It may have started from any block between the old finalized checkpoint and the new finalized checkpoint.

Reply:
This case is handled below.